### ISSUE TITLE
How to predict subtokens

### ISSUE THREAD

**Issue by chuanky** [tech-srl/slm-code-generation#16]
Dear authors,

I notice that in 3.3 Predicting Subtokens, you only describe how to copy a token. I was wondering how to generate a node such as 'EOStok' and 'EOSnode'?
In (Yin & Neubig, 2017), there is a probability that can control whether a token is copied or freely generated, but I did not find a similar mechanism here.

**Comment by urialon** [comment]
Hi @chuanky ,
Thank you for your interest in our work!

We predict AST nodes as described in the paragraph in Section 3.3 "Predicting AST nodes", which includes EOS_node.

EOS_token is predicted like a token, as described in "Predicting Subtokens" in the same subsection.

Differently from Yin & Neubig, we use the same joint softmax probability to both predict and copy, where some logits are computed using s_gen and some are computed using s_copy, as defined in Section 3.3.

I hope it helps!
Uri
